src/System.Management.Automation/engine/SessionStateLocationAPIs.cs

```diff
@@ -230,16 +230,29 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context)
             string driveName = null;
             ProviderInfo provider = null;
             string providerId = null;
+            bool pushNextLocation = true;
 
             // Replace path with last working directory when '-' was passed.
-            bool pushNextLocation = true;
             if (originalPath.Equals("-", StringComparison.OrdinalIgnoreCase))
             {
-                if (_SetLocationHistory.Count <= 0)
+                if (_SetLocationHistory.UndoCount <= 0)
                 {
-                    throw new InvalidOperationException(SessionStateStrings.SetContentToLastLocationWhenHistoryIsEmpty);
+                    throw new InvalidOperationException(SessionStateStrings.SetContentToLastLocationWhenHistoryUndoStackIsEmpty);
                 }
-                var previousLocation =  _SetLocationHistory.Pop();
+                var previousLocation =  _SetLocationHistory.Undo(this.CurrentLocation);
+                path = previousLocation.Path;
+                pushNextLocation = false;
+            }
+
+            // Replace path with last working directory from redo stack and push to the undo stack when '+' was passed.
+            if (originalPath.Equals("+", StringComparison.OrdinalIgnoreCase))
```

Comment:
We could use `switch` with C# 7.0 syntax and exclude `pushNextLocation`:
```c#
switch (originalPath)
{
    case var path when path.Equals("-", StringComparison.OrdinalIgnoreCase): 
        if (_SetLocationHistory.UndoCount <= 0) 
        { 
            throw new InvalidOperationException(SessionStateStrings.SetContentToLastLocationWhenHistoryUndoStackIsEmpty); 
        } 

        var previousLocation =  _SetLocationHistory.Undo(this.CurrentLocation); 
        path = previousLocation.Path; 
        break;
    case var path when path.Equals("+", StringComparison.OrdinalIgnoreCase);
        if (_SetLocationHistory.RedoCount <= 0) 
        { 
            throw new InvalidOperationException(SessionStateStrings.SetContentToLastLocationWhenHistoryRedoStackIsEmpty); 
        } 

        var previousLocation = _SetLocationHistory.Redo(); 
        _SetLocationHistory.Push(this.CurrentLocation); 
        path = previousLocation.Path; 
        break;
    default:
        var newPushPathInfo = GetNewPushPathInfo(); 
        _SetLocationHistory.Push(newPushPathInfo); 

        if (_SetLocationHistory.RedoCount >= 0) 
        { 
            _SetLocationHistory.InvalidateRedoStack(); 
        } 

        break;
}
```

Reply:
Yes, we could.